core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -985,10 +1012,12 @@ public Builder addSnapshot(Snapshot snapshot) {
         return this;
       }
 
+      ValidationException.check(!schemas.isEmpty(), "Attempting to add a snapshot before a schema is added");
+      ValidationException.check(!specs.isEmpty(), "Attempting to add a snapshot before a partition spec is added");
+      ValidationException.check(!sortOrders.isEmpty(), "Attempting to add a snapshot before a sort order is added");
```

Comment:
When building metadata from empty, the table specs aren't initialized. These assertions ensure they are set before adding files, so files added during a transaction line up with the table specs.